core/src/main/kotlin/com/justai/jaicf/BotEngine.kt

```diff
@@ -98,93 +96,67 @@ class BotEngine(
         reactions.loggingContext = loggingContext
         reactions.botContext = botContext
 
         processContext(botContext, requestContext)
 
         withHook(BotRequestHook(botContext, request, reactions)) {
-            val activation = botContext.isActiveSlotFilling().not()
-                .ifTrue { selectActivation(botContext, request) }
-                .withAppliedSlotFilling(botContext, request, reactions, cm)
-                .let { if (it.first) return else it.second }
-
-            loggingContext.activationContext = activation
-
-            if (activation?.activation == null) {
-                logger.warn("No state selected to handle a request $request")
-            } else {
-                val context = ProcessContext(
-                    request,
-                    reactions,
-                    requestContext,
-                    botContext,
-                    activation,
-                    loggingContext
-                )
-
-                processStates(context)
-                saveContext(cm, botContext, request, reactions)
-            }
-            conversationLoggers.forEach { it.obfuscateAndLog(loggingContext) }
-            botContext.cleanTempData()
+            processRequest(botContext, request, requestContext, reactions, loggingContext)
         }
+
+        botContext.cleanTempData()
+        saveContext(manager, botContext, request, reactions)
+        conversationLoggers.forEach { it.obfuscateAndLog(loggingContext) }
     }
 
-    private fun fillSlots(
-        selectedActivator: ActivationContext?,
+    private fun processRequest(
         botContext: BotContext,
         request: BotRequest,
+        requestContext: RequestContext,
         reactions: Reactions,
-        cm: BotContextManager
-    ): Pair<Boolean, ActivationContext?> {
-        val slotFillingActivatorName = botContext.getSlotFillingActivator()
-        val isSlotFillingSession = slotFillingActivatorName != null
-        var activationContext = selectedActivator
-        var shouldReturn = false
-
-        val res = when (val a = activationContext?.activator ?: getActivatorForName(slotFillingActivatorName)) {
-            null -> SlotFillingSkipped
-            else -> a.fillSlots(
-                request,
-                reactions,
-                botContext,
-                activationContext?.activation?.context,
-                slotReactor
-            )
-        }
-        if (res is SlotFillingInProgress) {
-            if (!isSlotFillingSession) {
-                botContext.startSlotFilling(activationContext?.activator?.name)
-                botContext.dialogContext.nextState = activationContext?.activation?.state
-                saveContext(cm, botContext, request, reactions)
+        loggingContext: LoggingContext
+    ) {
+        val slotFillingContext = if (isActiveSlotFilling(botContext)) {
+            getSlotFillingContext(botContext)!!
```

Comment:
I can change it, but for me, it looks more verbose with `if`

Reply:
Imo it can be rewritten to something like 
```kotlin
        val slotFillingContext = getSlotFillingCxtOrNull(botContext)
            ?: selectActivation(botContext, request)?.let { startSlotFilling(botContext, it) }
            ?: run { logger.warn("No state selected to handle a request $request"); return }
```

Reply:
@Denire, it can be rewritten like that, but it looks unclear why the first statement in any request processing is `getSlotFillingContext`.
`if (isActiveSlotFilling)` makes it more clear

Reply:
You can rename this function to `getSlotFillingContextIfActive`, and rewritten version will be still much easier to read then if/else